File: apps/quant-desktop/src/vfs/cache.ts
/**
 * Desktop LRU Chunk Cache
 *
 * Manages local cache of hydrated 64KB chunks with strict byte limit.
 */

export class LruChunkCacheTs {
  private maxBytes: number;
  private currentBytes = 0;
  private cache = new Map<string, Uint8Array>();

  constructor(maxBytes = 500 * 1024 * 1024) {
    // Default 500MB
    this.maxBytes = maxBytes;
  }

  public get(hash: string): Uint8Array | undefined {
    const data = this.cache.get(hash);
    if (!data) return undefined;

    // Refresh LRU order (delete & re-insert)
    this.cache.delete(hash);
    this.cache.set(hash, data);
    return data;
  }

  public put(hash: string, data: Uint8Array): void {
    const chunkLen = data.length;

    if (this.cache.has(hash)) {
      const existing = this.cache.get(hash)!;
      this.currentBytes = this.currentBytes - existing.length + chunkLen;
      this.cache.delete(hash);
      this.cache.set(hash, data);
      return;
    }

    // Evict oldest entries until within limit
    while (this.currentBytes + chunkLen > this.maxBytes && this.cache.size > 0) {
      const oldestKey = this.cache.keys().next().value;
      if (oldestKey) {
        const removed = this.cache.get(oldestKey)!;
        this.currentBytes -= removed.length;
        this.cache.delete(oldestKey);
      }
    }

    this.currentBytes += chunkLen;
    this.cache.set(hash, data);
  }

  public has(hash: string): boolean {
    return this.cache.has(hash);
  }

  public getCurrentBytes(): number {
    return this.currentBytes;
  }

  public size(): number {
    return this.cache.size;
  }

  public clear(): void {
    this.cache.clear();
    this.currentBytes = 0;
  }
}

File: apps/quant-desktop/src-tauri/src/vfs/cache.rs
//! Local LRU Chunk Cache
//!
//! Stores hydrated 64KB chunks in memory and local disk with strict byte limits (e.g. 10GB).
//! Evicts least-recently-used chunks when disk budget is exceeded.

use std::collections::{HashMap, VecDeque};
use std::sync::{Arc, Mutex};

pub struct LruChunkCache {
    max_bytes: u64,
    current_bytes: u64,
    cache: HashMap<String, Vec<u8>>,
    access_order: VecDeque<String>,
}

impl LruChunkCache {
    pub fn new(max_bytes: u64) -> Self {
        Self {
            max_bytes,
            current_bytes: 0,
            cache: HashMap::new(),
            access_order: VecDeque::new(),
        }
    }

    pub fn get(&mut self, hash: &str) -> Option<&Vec<u8>> {
        if self.cache.contains_key(hash) {
            // Move to end of access order
            self.access_order.retain(|h| h != hash);
            self.access_order.push_back(hash.to_string());
            self.cache.get(hash)
        } else {
            None
        }
    }

    pub fn put(&mut self, hash: String, data: Vec<u8>) {
        let chunk_len = data.len() as u64;

        // If chunk is already cached, update data and move to back
        if let Some(existing) = self.cache.get_mut(&hash) {
            self.current_bytes = self.current_bytes.saturating_sub(existing.len() as u64) + chunk_len;
            *existing = data;
            self.access_order.retain(|h| h != &hash);
            self.access_order.push_back(hash);
            return;
        }

        // Evict LRU chunks until within limit
        while self.current_bytes + chunk_len > self.max_bytes && !self.access_order.is_empty() {
            if let Some(lru_hash) = self.access_order.pop_front() {
                if let Some(removed) = self.cache.remove(&lru_hash) {
                    self.current_bytes = self.current_bytes.saturating_sub(removed.len() as u64);
                }
            }
        }

        self.current_bytes += chunk_len;
        self.cache.insert(hash.clone(), data);
        self.access_order.push_back(hash);
    }

    pub fn contains(&self, hash: &str) -> bool {
        self.cache.contains_key(hash)
    }

    pub fn current_bytes(&self) -> u64 {
        self.current_bytes
    }

    pub fn max_bytes(&self) -> u64 {
        self.max_bytes
    }

    pub fn len(&self) -> usize {
        self.cache.len()
    }

    pub fn is_empty(&self) -> bool {
        self.cache.is_empty()
    }
}
